File: autonav_ws/src/scr_core/src/node.cpp
#include "scr_core/node.h"
#include <unistd.h>

#define UNUSED(expr) do { (void)(expr); } while (0)

namespace SCR
{
	Node::Node(std::string node_name) : rclcpp::Node(node_name)
	{
		// State System
		deviceStates[node_name] = DeviceState::OFF;
		systemStateSubscriber = this->create_subscription<scr_msgs::msg::SystemState>("/scr/state/system", 100, std::bind(&Node::onSystemState, this, std::placeholders::_1));
		deviceStateSubscriber = this->create_subscription<scr_msgs::msg::DeviceState>("/scr/state/device", 100, std::bind(&Node::onDeviceState, this, std::placeholders::_1));
		resetSubscriber = this->create_subscription<std_msgs::msg::Empty>("/scr/reset", 100, std::bind(&Node::onResetInternal, this, std::placeholders::_1));
		deviceStateClient = this->create_client<scr_msgs::srv::SetDeviceState>("/scr/state/set_device_state");
		systemStateClient = this->create_client<scr_msgs::srv::SetSystemState>("/scr/state/set_system_state");
		resetPublisher = this->create_publisher<std_msgs::msg::Empty>("/scr/reset", 100);
		logPublisher = this->create_publisher<scr_msgs::msg::Log>("/scr/logging", 100);

		// Configuration
		auto configurationSubscriber = this->create_subscription<scr_msgs::msg::ConfigurationInstruction>("/scr/configuration", 100, std::bind(&Configuration::onConfigurationInstruction, &config, std::placeholders::_1));
		auto configurationLoadSubscriber = this->create_subscription<std_msgs::msg::String>("/scr/configuration/preset", 100, std::bind(&Configuration::onPresetChanged, &config, std::placeholders::_1));
		auto configurationPublisher = this->create_publisher<scr_msgs::msg::ConfigurationInstruction>("/scr/configuration", 100);
		auto configurationLoadPublisher = this->create_publisher<std_msgs::msg::String>("/scr/configuration/load", 100);
		config = Configuration(node_name, configurationSubscriber, configurationPublisher, configurationLoadSubscriber, configurationLoadPublisher);

		// Performance
		auto performancePublisher = this->create_publisher<scr_msgs::msg::PerformanceResult>("/scr/performance", 100);
		performance = Performance(node_name, performancePublisher);
	}

	Node::~Node()
	{
	}

	void Node::configure()
	{
	}

	void Node::onReset()
	{
	}

	void Node::onResetInternal(const std_msgs::msg::Empty::SharedPtr msg)
	{
		UNUSED(msg);
		onReset();
	}

	void Node::setEStop(bool state)
	{
		auto request = std::make_shared<scr_msgs::srv::SetSystemState::Request>();
		request->state = this->state.state;
		request->estop = state;
		request->mobility = this->state.mobility;
		request->mode = this->state.mode;
		systemStateClient->async_send_request(request);
	}

	void Node::setMobility(bool state)
	{
		auto request = std::make_shared<scr_msgs::srv::SetSystemState::Request>();
		request->state = this->state.state;
		request->estop = this->state.estop;
		request->mobility = state;
		request->mode = this->state.mode;
		systemStateClient->async_send_request(request);
	}

	void Node::setSystemState(SystemState state)
	{
		// Send a system state client request
		auto request = std::make_shared<scr_msgs::srv::SetSystemState::Request>();
		request->state = static_cast<uint8_t>(state);
		request->estop = this->state.estop;
		request->mobility = this->state.mobility;
		request->mode = this->state.mode;
		systemStateClient->async_send_request(request);
	}

	void Node::setSystemMode(SystemMode mode)
	{
		auto request = std::make_shared<scr_msgs::srv::SetSystemState::Request>();
		request->state = this->state.state;
		request->estop = this->state.estop;
		request->mobility = this->state.mobility;
		request->mode = static_cast<uint8_t>(mode);
		systemStateClient->async_send_request(request);
	}

	void Node::setDeviceState(DeviceState state)
	{
		auto request = std::make_shared<scr_msgs::srv::SetDeviceState::Request>();
		request->device = this->get_name();
		request->state = static_cast<uint8_t>(state);
		deviceStateClient->async_send_request(request);
	}

	void Node::log(const std::string& message)
	{
		auto msg = scr_msgs::msg::Log();
		msg.data = message;
		msg.node = get_name();
		logPublisher->publish(msg);
	}

	void Node::reset()
	{
		auto msg = std_msgs::msg::Empty();
		resetPublisher->publish(msg);
	}

	void Node::onSystemState(const scr_msgs::msg::SystemState::SharedPtr msg)
	{
		if(msg->state == SystemState::SHUTDOWN)
		{
			kill(getpid(), SIGINT);
		}

		auto old = state;
		state = *msg;
		transition(old, state);
	}

	void Node::onDeviceState(const scr_msgs::msg::DeviceState::SharedPtr msg)
	{
		bool isNew = deviceStates.find(msg->device) == deviceStates.end();
		deviceStates[msg->device] = static_cast<DeviceState>(msg->state);
		onDeviceStateUpdated(msg, isNew && msg->state != SCR::DeviceState::OFF);
		if (msg->device != this->get_name())
		{
			return;
		}

		if (msg->state == static_cast<uint8_t>(DeviceState::STANDBY))
		{
			config.recache();
			configure();
			return;
		}
	}

	void Node::transition(scr_msgs::msg::SystemState old, scr_msgs::msg::SystemState updated)
	{
		UNUSED(old);
		UNUSED(updated);
		throw std::runtime_error("Transition function not overridden");
	}
	
	void Node::onDeviceStateUpdated(scr_msgs::msg::DeviceState::SharedPtr msg, bool isNew)
	{
	}

	scr_msgs::msg::SystemState Node::getSystemState()
	{
		return state;
	}

	DeviceState Node::getDeviceState()
	{
		return deviceStates[this->get_name()];
	}

	DeviceState Node::getDeviceState(std::string id)
	{
		return deviceStates[id];
	}

	std::map<std::string, DeviceState> Node::getDeviceStates()
	{
		return deviceStates;
	}
}

File: autonav_ws/src/scr_core/scr_core/node.py
from scr_core.state import DeviceStateEnum, SystemStateEnum
from scr_msgs.srv import SetDeviceState, SetSystemState
from scr_msgs.msg import DeviceState, SystemState, Log
from scr_core.configuration import Configuration
from scr_core.performance import Performance
from rclpy.node import Node as ROSNode
from std_msgs.msg import Empty
import time
import signal
import os


class Node(ROSNode):
    """
    A Node in the ROS graph.

    Specialized for SCR (Sooner Competitive Robotics), includes systems for state and configuration.
    """

    def __init__(self, node_name):
        super().__init__(node_name)
        self.id = node_name

        # State System
        self.config = Configuration(node_name, self)
        self.deviceStateSubscriber = self.create_subscription(DeviceState, "/scr/state/device", self.onDeviceState, 100)
        self.systemStateSubscriber = self.create_subscription(SystemState, "/scr/state/system", self.onSystemState, 100)
        self.deviceStateClient = self.create_client(SetDeviceState, "/scr/state/set_device_state")
        self.systemStateClient = self.create_client(SetSystemState, "/scr/state/set_system_state")
        self.resetSubscriber = self.create_subscription(Empty, "/scr/reset", self.onResetInternal, 100)
        self.resetPublisher = self.create_publisher(Empty, "/scr/reset", 100)
        self.logPublisher = self.create_publisher(Log, "/scr/logging", 100)

        # Configuration
        self.config = Configuration(self.id, self)
        self.deviceStates = {}
        self.state = SystemState()

        # Performance
        self.performance = Performance(self)

    def configure(self):
        pass

    def onReset(self):
        pass

    def onResetInternal(self, _):
        self.onReset()

    def reset(self):
        self.resetPublisher.publish(Empty())

    def getDeviceID(self) -> int:
        return self.id
    
    def log(self, message: str):
        log = Log()
        log.node = self.get_name()
        log.data = message
        self.logPublisher.publish(log)

    def setSystemStateInternal(self, state: SystemState):
        request = SetSystemState.Request()
        request.state = state.state
        request.estop = state.estop
        request.mobility = state.mobility
        self.systemStateClient.call_async(request)

    def setDeviceState(self, state: DeviceStateEnum):
        request = SetDeviceState.Request()
        request.state = state
        request.device = self.id
        self.deviceStateClient.call_async(request)

    def setSystemState(self, state: SystemStateEnum):
        self.setSystemStateInternal(SystemState(
            state = state,
            estop = self.state.estop,
            mobility = self.state.mobility
        ))
    
    def setEStop(self, state: bool):
        self.setSystemStateInternal(SystemState(
            state = self.state.state,
            estop = state,
            mobility = self.state.mobility
        ))

    def setMobility(self, state: bool):
        self.setSystemStateInternal(SystemState(
            state = self.state.state,
            estop = self.state.estop,
            mobility = state
        ))

    def onSystemState(self, state: SystemState):
        if state.state == SystemStateEnum.SHUTDOWN:
            os.kill(os.getpid(), signal.SIGINT)

        old = self.state
        self.state = state

        if self.getDeviceState() == DeviceStateEnum.OFF:
            return

        self.transition(old, state)

    def getClockNs(self) -> int:
        return time.time() * 1000000
        
    def getClockMs(self) -> int:
        return time.time() * 1000
    
    def getClockSec(self) -> float:
        return time.time()

    def onDeviceState(self, state: DeviceState):
        self.deviceStates[state.device] = state.state
        if state.device != self.id:
            return
        
        if state.state == DeviceStateEnum.STANDBY:
            self.config.recache()
            self.configure()
            self.onSystemState(self.state)

    def transition(self, old: SystemState, updated: SystemState):
        raise NotImplementedError()

    def getSystemState(self) -> SystemState:
        return self.state

    def getDeviceState(self, id: str = None) -> DeviceStateEnum:
        if id is None:
            return DeviceStateEnum.OFF if self.id not in self.deviceStates else self.deviceStates[self.id]
        return self.deviceStates[self.id] if id is None else self.deviceStates[self.id]
    
    def getDeviceStates(self):
        return self.deviceStates
